core/state/statedb.go

```diff
@@ -837,12 +846,22 @@ func (s *StateDB) IntermediateRoot(deleteEmptyObjects bool) common.Hash {
 
 	// If there was a trie prefetcher operating, terminate it async so that the
 	// individual storage tries can be updated as soon as the disk load finishes.
-	if s.prefetcher != nil {
-		s.prefetcher.terminate(true)
+	if s.collectWitness && s.prefetcher != nil {
```

Comment:
Looks like you can keep `if s.prefetcher != nil` as the outer clause, for simplicity